src/arithmetic.jl

```diff
@@ -1,90 +1,78 @@
+    circshift(sa::SecureArray{<:AbstractCryptoBackend, N}, shift::NTuple{N, Integer})
 
-Circularly shift, i.e., rotate the data in `sv` by `shift` positions, similarly to Julia's
-`circshift` for regular arrays. `wrap_by` indicates whether the rotation should be applied
-with respect to the current data length of `sv` (`wrap_by :length`) or with respect to its
-maximum capacity (`wrap_by = :capacity`).
+Circularly shift, i.e., rotate the data in `sa` by `shift` positions, similarly to Julia's
+`circshift` for regular arrays.
 
-Note: If `sv`'s length is less than its capacity, wrapping by `:length` increases the
-multiplicative depth of your algorithm by one and is more expensive to compute. Furthermore,
-one additional rotation is applied with a shift of
-`-sign(shift) * (length(sv) - abs(shift))`.
+Note: If `N` is greater than one, this operation increases the multiplicative level by two.
 
-See also: [`SecureVector`](@ref), [`length`](@ref), [`capacity`](@ref)
-"""
-function Base.circshift(sv::SecureVector, shift::Integer; wrap_by = :capacity)
-    if !(wrap_by in (:length, :capacity))
-        throw(ArgumentError("Unsupported value '$wrap_by' passed to `wrap_by` (must be `:length` or `:capacity`)"))
-    end
+Note: To precompute all required rotation indexes, use `init_array_rotation!`.
 
-    if shift == 0
-        return sv
+See also: [`SecureArray`](@ref), [`init_array_rotation!`](@ref)
+"""
+function Base.circshift(sa::SecureArray{<:AbstractCryptoBackend, N}, shift::NTuple{N, Integer}) where N
+    if all(shift .% size(sa) .== 0)
+        return sa
     end
 
-    rotate(sv, shift; wrap_by)
+    rotate(sa, shift)
 end
 
+"""
+    circshift(sv::SecureVector, shift::Union{Integer, Tuple{Integer}}; wrap_by = :not_set)
 
-############################################################################################
-# Matrix
-############################################################################################
-
-# Add
-Base.:+(sm1::SecureMatrix, sm2::SecureMatrix) = add(sm1, sm2)
-Base.:+(sm::SecureMatrix, pm::PlainMatrix) = add(sm, pm)
-Base.:+(pm::PlainMatrix, sm::SecureMatrix) = add(sm, pm)
-Base.:+(sm::SecureMatrix, scalar::Real) = add(sm, scalar)
-Base.:+(scalar::Real, sm::SecureMatrix) = add(sm, scalar)
-
-# Subtract
-Base.:-(sm1::SecureMatrix, sm2::SecureMatrix) = subtract(sm1, sm2)
-Base.:-(sm::SecureMatrix, pm::PlainMatrix) = subtract(sm, pm)
-Base.:-(pm::PlainMatrix, sm::SecureMatrix) = subtract(pm, sm)
-Base.:-(sm::SecureMatrix, scalar::Real) = subtract(sm, scalar)
-Base.:-(scalar::Real, sm::SecureMatrix) = subtract(scalar, sm)
+Circularly shift, i.e., rotate the data in `sv` by `shift` positions, similarly to Julia's
+`circshift` for regular arrays. `wrap_by` indicates whether the rotation should be applied
+with respect to the current data length of `sv` (`wrap_by :length`) or with respect to its
+maximum capacity (`wrap_by = :capacity`). Default wrap_by = :not_set results in 
```

Comment:
```suggestion
maximum capacity (`wrap_by = :capacity`). Default wrap_by = :auto results in 
```
I would call the default just `:default` or `:auto`. However, in light of our focus on usability, shouldn't we make wrapping by length the default in all cases? I mean, nobody except for very special (performance-related) circumstances wants to circshift by capacity, right?

And we should definitely *not* differentiate between having one or more internal ciphertexts. That's an implementation detail the user should **not** need to worry about

Reply:
Okay, so since `:capacity` makes sense only for 1d and single ciphertext, that is implementation detail, I would then completely remove `wrap_by` argument. If someone wants then `:capacity`, can use `OpenFHE.EvalRotate`, correct?

Reply:
Completed

I would also then replace `init_rotation!` with `init_array_rotation!` because otherwise, a user should precompute all required rotation indexes (since now we are only ise `wrap_by = :length`) himself looking into code implementation or using `init_array_rotation!`. Note that now required indexes differ from those that were before for `wrap_by = :length`. This will of course require breaking release, but since required indexes are already different, it will be breaking any way without adapting `circshift` algorithm to avoid this. What do you think?

Reply:
> If someone wants then `:capacity`, can use `OpenFHE.EvalRotate`, correct?

Correct 👍

> I would also then replace `init_rotation!` with `init_array_rotation!`

I do not understand. Do we still need to "init rotation" functions at all? IMHO there should be just one user-facing function called `init_rotation!`, which should do the Right Thing and accept a tuple (or multiple) of indices that are identical to those that would be passed to `circshift` later. But maybe I misunderstand something here

Reply:
Done